osu.Game/Screens/Edit/Timing/TapTimingControl.cs

```diff
@@ -223,7 +260,21 @@ private void adjustBpm(double adjust)
             if (timing == null)
                 return;
 
-            timing.BeatLength = 60000 / (timing.BPM + adjust);
+            double newBeatLength = 60000 / (timing.BPM + adjust);
+
+            List<HitObject> hitObjectsInRange = hitObjectsInTimingRange(beatmap, selectedGroup.Value!);
+
+            if (adjustPlacedNotes.Current.Value)
+            {
+                foreach (HitObject hitObject in hitObjectsInRange)
+                {
+                    double beat = (hitObject.StartTime - selectedGroup.Value!.Time) / timing.BeatLength;
+
+                    hitObject.StartTime = beat * newBeatLength + selectedGroup.Value.Time;
```

Comment:
The tooltip for this says that it will also handle BPM but how does it do that? It only shifts the objects around. For offset sure that is enough, but it will not be enough for BPM won't it?

Counterexample: sliders - changing BPM for a slider snapped to N beats means that the slider duration should also change, but this doesn't do anything about that. Which means that the slider will become unsnapped after BPM adjustment. And as far as I can tell the stable implementation doesn't seem to do anything about that either.

So I kinda question as to whether this feature is "fully baked" in this form, so to speak.

Reply:
Shifting around was sufficient for HitCircles, and somewhat for sliders (being correct when using undo/redo, just not at the moment of change - maybe due to something in the update loop of a slider? haven't looked into it yet)

But yeah, thanks for pointing out Spinners, forgot to check their behaviour. I'll change it to also adjust their length.

Reply:
Spinners (and other mode equivalents) should be addressed now. 

Although am still having trouble getting sliders to work without some glitchy mechanics... Currently every slider except for those on the timing point work.
May be to do with snapping, the hitobjects landing still in range for the previous timing point even though they're a rounding error away from the would-be correct timing point. Not sure yet.

Reply:
I've figured out what causes this - is just because the change doesn't modify the bindable value when they're the same.

I'm not sure how to exactly resolve this issue, as I can't see any non jank way of triggering a change. Obviously just adding and removing some constant would work, but I don't particularly want to implement that solution.
```
double beat = (hitObject.StartTime - selectedGroup.Value!.Time) / timing.BeatLength;

hitObject.StartTime = (beat * newBeatLength) + selectedGroup.Value.Time;
hitObject.StartTime += 1;
hitObject.StartTime -= 1;
```

I had hopes that calling `hitObject.StartTimeBindable.TriggerChange();` would work, but this does not propagate to bindables bound to this one.

I believe it's this bindable that is causing the issue, which mostly comes down to how `HitObject.ApplyDefaultsToSelf` works.
https://github.com/ppy/osu/blob/058d7aa7c1ae14fb998413785d50a2d6b5384433/osu.Game/Rulesets/Edit/PlacementBlueprint.cs#L81-L86

An idea is making all such calculations for sliders instead be calculated in getters, which should be fairly doable. Just would need to implement a cache of the latest `ControlPointInfo` and `IBeatmapDifficultyInfo` call.

Although, I guess I don't see any reason why I can't just call `hitObject.ApplyDefaults(beatmap.ControlPointInfo, beatmap.Difficulty);` after changing `timing.BeatLength` when adjusting BPM. Feels somewhat iffy a solution to me, but from testing does work.

Going to rest for a bit, but let me know if you think I should implement a specific solution @bdach. Otherwise, I'll probably go for what I mentioned with changing the getters - I've seen issues with calculations done in `ApplyDefaults` not being applied fairly often, I feel, so I think it's best to move calculations out of there when possible.

Reply:
I've decided not to go via modifying the getters for now, and instead created a discussion #28402 about replacing all instances of `ApplyDefaultsToSelf` with just modifications of the getter and setter.

For the moment, I'll just do this by calling `hitObject.ApplyDefaults(beatmap.ControlPointInfo, beatmap.Difficulty);` instead.

Reply:
Please revert https://github.com/ppy/osu/pull/28382/commits/3d5a04ac99d4b9f2002a24206106472c53da5c7b and apply this instead:

```diff
diff --git a/osu.Game/Screens/Edit/Timing/TapTimingControl.cs b/osu.Game/Screens/Edit/Timing/TapTimingControl.cs
index cffa31d117..f74b7ac079 100644
--- a/osu.Game/Screens/Edit/Timing/TapTimingControl.cs
+++ b/osu.Game/Screens/Edit/Timing/TapTimingControl.cs
@@ -275,6 +275,8 @@ private void adjustBpm(double adjust)
 
                     if (hitObject is not IHasRepeats && hitObject is IHasDuration hitObjectWithDuration)
                         hitObjectWithDuration.Duration *= newBeatLength / timing.BeatLength;
+
+                    beatmap.Update(hitObject);
                 }
             }
 

```

Yes this ideally should not be required but it is a far preferable workaround to yours because it uses established patterns. Start time changing does not need this because `StartTimeBindable` is hooked up to automatically call this.

Reply:
Ah, didn't realize that function exists... Before doing that, I'd still suggest that this call is done in the same location as i did in 3d5a04a, however.

This due to the behaviour i mentioned in the discussion post that exists on master currently, with sliders not changing length when control points change.

Reply:
Sure I suppose.